From dc1b670a2c34a0e1c5c0ba9531dfc8f5a6746a0b Mon Sep 17 00:00:00 2001
From: Ronald S. Bultje <rsbultje@gmail.com>
Date: Sun, 10 Jul 2011 21:23:09 -0700
Subject: [PATCH 53/57] vp8/mt: flush worker thread, not application thread context, on seek.

This prevents a crash when seeking.
(cherry picked from commit d1cf45911935cc4fed9afd3a37d99616d31eb9da)
---
 libavcodec/pthread.c |    9 ++++++---
 libavcodec/utils.c   |    2 +-
 2 files changed, 7 insertions(+), 4 deletions(-)

diff --git a/libavcodec/pthread.c b/libavcodec/pthread.c
index 08ef4ba..9fea9a0 100644
--- a/libavcodec/pthread.c
+++ b/libavcodec/pthread.c
@@ -746,9 +746,12 @@ void ff_thread_flush(AVCodecContext *avctx)
     if (!avctx->thread_opaque) return;
 
     park_frame_worker_threads(fctx, avctx->thread_count);
-
-    if (fctx->prev_thread)
-        update_context_from_thread(fctx->threads->avctx, fctx->prev_thread->avctx, 0);
+    if (fctx->prev_thread) {
+        if (fctx->prev_thread != &fctx->threads[0])
+            update_context_from_thread(fctx->threads[0].avctx, fctx->prev_thread->avctx, 0);
+        if (avctx->codec->flush)
+            avctx->codec->flush(fctx->threads[0].avctx);
+    }
 
     fctx->next_decoding = fctx->next_finished = 0;
     fctx->delaying = 1;
diff --git a/libavcodec/utils.c b/libavcodec/utils.c
index c32fda2..bbed726 100644
--- a/libavcodec/utils.c
+++ b/libavcodec/utils.c
@@ -1058,7 +1058,7 @@ void avcodec_flush_buffers(AVCodecContext *avctx)
 {
     if(HAVE_PTHREADS && avctx->active_thread_type&FF_THREAD_FRAME)
         ff_thread_flush(avctx);
-    if(avctx->codec->flush)
+    else if(avctx->codec->flush)
         avctx->codec->flush(avctx);
 }
 
-- 
1.7.4.1

